### ISSUE TITLE
Feature: Multiple validation rules for the same name/type

### ISSUE THREAD

**Issue by kettanaito** [kettanaito/react-advanced-form#114]
**Branch:** [`feature/validation-multiple-rules`](https://github.com/kettanaito/react-advanced-form/tree/feature/validation-multiple-rules)

## What
Need to have a support for multiple validation rules for the dedicated name or type.

## Why
This would grant a flexible validation procedure coupled with the respective validation messages.

## How
Let the validation groups have the named validation rules. Those names will be then used in the validation messages to bind them to the respective message.

**Validation rules:**
```js
{
  type: {
    password: {
      minLength: ({ value }) => (value.length > 7),
      capitalLetter: ({ value }) => ...
    }
  },
  name: {
    confirmPassword: {
      matches: ({ value, fields }) => (value === fields.password.value)
    }
  }
}
```

**Validation messages:**
```js
{
  type: {
    password: {
      missing: 'Please provide the password.',
      invalid: 'General message for unexpected field; fallback for the named rules without the corresponding named message.',
      custom/specific/byRule: {
        minLength: 'Password must be at least 8 characters long.',
        capitalLetter: 'Include at least one capital letter.'
      }
    }
  },
  name: {
    confirmPassword: {
      matches: 'The passwords do not match.'
    }
  }
}
```

This should validate the `<Field.Input type="password" name="confirmPassword" />` relatively to the password "type" and its messages, and to the "confirmPassword" name and its messages.

**Comment by kettanaito** [comment]
I guess that a good way to organize those multiple rules would be to store them under one agreed property name, let's say
```
rules.type[fieldType].custom[customRuleName]
```

### What definitely should be:
* **Maintainable API**. In case of API change it shouldn't take a lot of changes to perform during the migration. Having them grouped under one defined property name ensures only one place needs to change, potentially.
* **Intuitive**. There should be a clean separation of default messages `missing`/`invalid` and the custom ones.


What do you think, @Vidlec?

**Comment by Vidlec** [comment]
Interesting. Looks good, i like the idea of grouping the messages... 

As for the passed rules... I am not sure how to handle this at all :)
Regarding the order of validations... Lets just let it validate from top to bottom and then if someone need different order its his responsibility to structure config properly.

If we go the other way - validating everything, this concern gone. But ofc. it can be non-performing.

**Comment by Vidlec** [comment]
And I agree messages and rules should be in different files.

**Comment by kettanaito** [comment]
Changes are going to be implemented on [feature/validation-refactoring](https://github.com/kettanaito/react-advanced-form/tree/feature/validation-refactoring) branch.

**Comment by kettanaito** [comment]
Get the path stamp of the validation *rule* and apply it to the validation messages Map directly.

```
const ruleStamp = ['type', 'text']; // ==> { type: { text: f() } }
['name', 'username', 'customRule']; // ==> { name: { username: { customRule: f() } } } 
```

**Comment by kettanaito** [comment]
## Update
* In case the rule passes, there will be a `valid` message property. It behaves the same way as other message properties.
* It will be possible to declare multiple name-specific and type-specific rules
* The sequence of validation remains the same: `name > type`
* In case of multiple name-specific rules, *all* are executed, but remaining type-specific rules are *ignored*

**Comment by kettanaito** [comment]
## TODO
- [ ] Provide the logic for this scenario:
```js
name: {
  fieldName: {
    firstRule: () => {},
    secondRule: () => {}
  }
}
```

* If both `firstRule` and `secondRule` have error messages, and both rules fail, display both error messages
* If `firstRule` has error message, but `secondRule` doesn't, display only the error message for `firstRule`.
* If `secondRule` is *the only* rule, and it doesn't have an error message, display fallback error message

**Comment by kettanaito** [comment]
## Update
Current change on the feature branch is quite good, yet there is one moment I do not like.

When multiple named rules are complementary, displaying multiple named error messages doesn't make much sense. Consider:

```js
// validation-rules.js
{
  name: {
    fieldName: {
      numbersOnly: ({ value }) => /^\d+$/.test(value),
      maxSize: ({ value }) => (value < 10)
    }
  }
}

// validation-messages.js
{
  name: {
    fieldName: {
      rules: {
        numbersOnly: 'Please provide numbers only',
        maxSize: 'Should be less than 10'
      }
    }
  }
}
```

Now, if passing a "foo" as a value to "fieldName", both validation messages for both rules are shown:

```
Please provide numbers only
Should be less than 10
```

This doesn't make sense to me. Multiple named messages work great if their rules point into different directions (`capitalLetter`, `atLeastOneNumber`, etc.).

**Comment by kettanaito** [comment]
Changes implemented and are available since version [`0.18.0`](https://github.com/kettanaito/react-advanced-form/commit/31f13d68a5983c91ff8bf5f4ff1626a66cf69af3). Published.

Closing.
